File: src/rosterizer/Roster.java
package rosterizer;

import net.bytebuddy.dynamic.scaffold.MethodGraph;

import javax.swing.*;
import javax.swing.filechooser.FileSystemView;
import java.io.File;
import java.io.FileNotFoundException;
import java.util.*;

public class Roster {
    private HashMap<Integer, LinkedList<String>> assignments;
    private HashMap<String, Associate> associates;
    private AssociateDB database;

    private int problemSolverMax = 5;
    private int waterspiderMax = 3;
    private int eolMax = 3;
    private int refurbMax = 2;
    private int whdMax = 1;
    private int icqaMax = 1;
    private int recoveryMax = 1;
    private int shoeProcessingMax = 8;
    

    public Roster(AssociateDB database) {
        assignments = new HashMap<>();
        associates = new HashMap<>();

        this.database = database;
    }

    public void setAssignments() {
        LinkedList<Associate> unassignedAssociates = new LinkedList<>(associates.values());

        Collections.shuffle(unassignedAssociates);

        // Selecting problem solvers
        assignments.put(Associate.Role.PS.value, new LinkedList<>());
        System.out.println("---PROBLEM SOLVERS---");
        for(int i = 0; assignments.get(Associate.Role.PS.value).size() < problemSolverMax; i++)
            if(unassignedAssociates.get(i).hasRole(Associate.Role.PS.value)) {
                assignments.get(Associate.Role.PS.value).add(unassignedAssociates.get(i).getUsername());
                System.out.println(unassignedAssociates.get(i).getUsername());
                unassignedAssociates.remove(i);
            }
        // Selecting waterspiders
        assignments.put(Associate.Role.WATERSPIDER.value, new LinkedList<>());
        System.out.println("---WATERSPIDERS---");
        for(int i = 0; assignments.get(Associate.Role.WATERSPIDER.value).size() < waterspiderMax; i++)
            if(unassignedAssociates.get(i).hasRole(Associate.Role.WATERSPIDER.value)) {
                assignments.get(Associate.Role.WATERSPIDER.value).add(unassignedAssociates.get(i).getUsername());
                System.out.println(unassignedAssociates.get(i).getUsername());
                unassignedAssociates.remove(i);
            }
        // Selecting eol
        assignments.put(Associate.Role.EOL.value, new LinkedList<>());
        System.out.println("---EOL---");
        for(int i = 0; assignments.get(Associate.Role.EOL.value).size() < eolMax; i++)
            if(unassignedAssociates.get(i).hasRole(Associate.Role.EOL.value)) {
                assignments.get(Associate.Role.EOL.value).add(unassignedAssociates.get(i).getUsername());
                System.out.println(unassignedAssociates.get(i).getUsername());
                unassignedAssociates.remove(i);
            }
        // Selecting whd
        assignments.put(Associate.Role.WHD.value, new LinkedList<>());
        System.out.println("---WHD---");
        for(int i = 0; assignments.get(Associate.Role.WHD.value).size() < whdMax; i++)
            if(unassignedAssociates.get(i).hasRole(Associate.Role.WHD.value)) {
                assignments.get(Associate.Role.WHD.value).add(unassignedAssociates.get(i).getUsername());
                System.out.println(unassignedAssociates.get(i).getUsername());
                unassignedAssociates.remove(i);
            }
        // Selecting refurb
        assignments.put(Associate.Role.REFURB.value, new LinkedList<>());
        System.out.println("---REFURB---");
        for(int i = 0; assignments.get(Associate.Role.REFURB.value).size() < refurbMax; i++)
            if(unassignedAssociates.get(i).hasRole(Associate.Role.REFURB.value)) {
                assignments.get(Associate.Role.REFURB.value).add(unassignedAssociates.get(i).getUsername());
                System.out.println(unassignedAssociates.get(i).getUsername());
                unassignedAssociates.remove(i);
            }
        // Selecting icqa
        assignments.put(Associate.Role.ICQA.value, new LinkedList<>());
        System.out.println("---ICQA---");
        for(int i = 0; assignments.get(Associate.Role.ICQA.value).size() < icqaMax; i++)
            if(unassignedAssociates.get(i).hasRole(Associate.Role.ICQA.value)) {
                assignments.get(Associate.Role.ICQA.value).add(unassignedAssociates.get(i).getUsername());
                System.out.println(unassignedAssociates.get(i).getUsername());
                unassignedAssociates.remove(i);
            }
        // Selecting recovery
        assignments.put(Associate.Role.RECOVERY.value, new LinkedList<>());
        System.out.println("---RECOVERY---");
        for(int i = 0; assignments.get(Associate.Role.RECOVERY.value).size() < recoveryMax; i++)
            if(unassignedAssociates.get(i).hasRole(Associate.Role.RECOVERY.value)) {
                assignments.get(Associate.Role.RECOVERY.value).add(unassignedAssociates.get(i).getUsername());
                System.out.println(unassignedAssociates.get(i).getUsername());
                unassignedAssociates.remove(i);
            }
        // Selecting shoe processing
        assignments.put(Associate.Role.SHOES.value, new LinkedList<>());
        System.out.println("---SHOE PROCESSING---");
        for(int i = 0; assignments.get(Associate.Role.SHOES.value).size() < shoeProcessingMax; i++)
            if(unassignedAssociates.get(i).hasRole(Associate.Role.SHOES.value)) {
                assignments.get(Associate.Role.SHOES.value).add(unassignedAssociates.get(i).getUsername());
                System.out.println(unassignedAssociates.get(i).getUsername());
                unassignedAssociates.remove(i);
            }
        // Selecting softlines processors
        assignments.put(Associate.Role.PROCESS.value, new LinkedList<>());
        System.out.println("---SOFTLINES PROCESSING---");
        for (Associate unassignedAssociate : unassignedAssociates) {
            assignments.get(Associate.Role.PROCESS.value).add(unassignedAssociate.getUsername());
            System.out.println(unassignedAssociate.getUsername());
        }
    }

    public void importSSPOT() {
        JFileChooser chooser = new JFileChooser(FileSystemView.getFileSystemView());
        LinkedList<String> rows = new LinkedList<>();
        String filepath = "";

        chooser.setMultiSelectionEnabled(true);
        int option = chooser.showOpenDialog(null);

        if(option == JFileChooser.APPROVE_OPTION)
            filepath = chooser.getSelectedFile().getAbsolutePath();

        try {
            File[] files = chooser.getSelectedFiles();

            for (File file : files) {
                Scanner sc = new Scanner(file);
                while (sc.hasNextLine())
                    rows.add(sc.nextLine());
                sc.close();
            }
        } catch(FileNotFoundException e) {
            e.printStackTrace();
            return;
        }

        rows.removeFirst();

        for(int i = 0; i < rows.size(); i++) {
            String[] rowAsArray = rows.get(i).replace("\"", "").split(",");
            Associate assoc = database.getAssociate(rowAsArray[1]);
            if(assoc != null)
                associates.put(assoc.getUsername(), assoc);
            else
                associates.put(rowAsArray[1], new Associate(rowAsArray[1]));
        }
    }

    @Override
    public String toString() {
        String result = "";
        return result;
    }
}

File: src/rosterizer/Roster.kt
package rosterizer

import java.io.File
import java.io.FileNotFoundException
import java.util.LinkedList
import java.util.Scanner

class Roster(private val processAssistants: List<String>) {
    private val scheduledAssociates = HashMap<String, Associate>()
    private val assignedRoles = HashMap<Int, LinkedList<Associate>>()

    fun importSSPOT(filepath: String = "") {
        if(filepath.isEmpty()) return

        val rows = LinkedList<String>()

        try {
            val file = File(filepath)
            val scanner = Scanner(file)

            while(scanner.hasNextLine())
                rows.add(scanner.nextLine())
            scanner.close()
        } catch(exception: FileNotFoundException) {
            println("FileNotFoundException thrown while importing SSPOT roster")
            return
        }

        rows.removeFirst()
        for(row in rows) {
            val assoc = Associate(row.split(",")[1].replace("\"", ""))

            // Filtering out process assistants from SSPOT roster
            if(processAssistants.contains(assoc.login))
                continue
            scheduledAssociates[assoc.login] = assoc
            //println("${scheduledAssociates.size}: ${assoc.login}")
        }
    }

    fun importTrainingQUIP(filepath: String = "") {
        if(filepath.isEmpty()) return

        val rows = LinkedList<String>()

        try {
            val file = File(filepath)
            val scanner = Scanner(file)

            while(scanner.hasNextLine())
                rows.add(scanner.nextLine())
            scanner.close()
        } catch(exception: FileNotFoundException) {
            println("FileNotFoundException thrown while importing Training QUIP")
            return
        }

        rows.removeFirst()
        for(row in rows) {
            val rowAsArray = row.substring(row.lastIndexOf('"') + 1, row.length - 1).split(",")
            val loginColValue: String = rowAsArray[1]
            val processTrained: String = rowAsArray[6].lowercase()

            if (scheduledAssociates.containsKey(loginColValue)) {
                when (processTrained) {
                    "ambassador" -> scheduledAssociates[loginColValue]!!.addTrainedRole(Associate.ROLE_AMBASSADOR)
                    "audit" -> scheduledAssociates[loginColValue]!!.addTrainedRole(Associate.ROLE_RECOVERY)
                    "end of line" -> scheduledAssociates[loginColValue]!!.addTrainedRole(Associate.ROLE_EOL)
                    "outbound" -> scheduledAssociates[loginColValue]!!.addTrainedRole(Associate.ROLE_OUTBOUND)
                    "pit" -> scheduledAssociates[loginColValue]!!.addTrainedRole(Associate.ROLE_PIT)
                    "problem solve" -> scheduledAssociates[loginColValue]!!.addTrainedRole(Associate.ROLE_PS)
                    "refurb" -> scheduledAssociates[loginColValue]!!.addTrainedRole(Associate.ROLE_REFURB)
                    "shoe icqa" -> scheduledAssociates[loginColValue]!!.addTrainedRole(Associate.ROLE_ICQA)
                    "tdr" -> scheduledAssociates[loginColValue]!!.addTrainedRole(Associate.ROLE_TDR)
                    "water spider" -> scheduledAssociates[loginColValue]!!.addTrainedRole(Associate.ROLE_WATERSPIDER)
                    "whd" -> scheduledAssociates[loginColValue]!!.addTrainedRole(Associate.ROLE_WHD)
                }
                scheduledAssociates[loginColValue]!!.addTrainedRole(Associate.ROLE_PROCESS)
            }
        }
    }

    fun randomizeRoles() {
        val associateList = scheduledAssociates.keys

        fun assignIndirectRole(role: Int, amount: Int, msg: String) {
            assignedRoles[role] = LinkedList()
            println("===== $msg =====")
            while(assignedRoles[role]!!.size < amount) {
                val rand = (0..<associateList.size).random()

                if(scheduledAssociates[associateList.elementAt(rand)]!!.hasRole(role)) {
                    assignedRoles[role]!!.add(scheduledAssociates[associateList.elementAt(rand)]!!)
                    println("-> ${associateList.elementAt(rand)}")
                    associateList.remove(associateList.elementAt(rand))
                }
            }
        }

        // Assigning indirect roles
        assignIndirectRole(Associate.ROLE_PS, 5, "Problem Solvers")
        assignIndirectRole(Associate.ROLE_WATERSPIDER, 4, "Waterspiders")
        assignIndirectRole(Associate.ROLE_OUTBOUND, 1, "Outbound")
        assignIndirectRole(Associate.ROLE_EOL, 3, "EOL")
        assignIndirectRole(Associate.ROLE_ICQA, 1, "ICQA")
        assignIndirectRole(Associate.ROLE_RECOVERY, 1, "Recovery")
        assignIndirectRole(Associate.ROLE_REFURB, 2, "Refurb")
        assignIndirectRole(Associate.ROLE_WHD, 1, "Warehouse Deals")

        // Assigning whatever associates remain to direct process
        assignedRoles[Associate.ROLE_PROCESS] = LinkedList()
        println("===== Processors =====")
        for(login in associateList) {
            assignedRoles[Associate.ROLE_PROCESS]!!.add(scheduledAssociates[login]!!)
            println("-> $login")
        }
    }

    fun getAssociateLoginsByRole(role: Int): List<String>? = assignedRoles[role]?.map{it.login}

    override fun toString(): String {
        val result = "Problem Solve,,Waterspider,,Outbound,,EOL,,ICQA,,Recovery,,Refurb,,WHD\n"

        var i = 0
        while(i < assignedRoles[Associate.ROLE_PROCESS]!!.size) {

        }

        return result
    }
}